### ISSUE TITLE
MAX_CURVELIST too small

### ISSUE THREAD

**Issue by tycho** [openssl/openssl#5232]
The value for `MAX_CURVELIST` of `28` in `t1_lib.c` is too small to support listing all the available curves.

This works:
> $ openssl s_client -curves sect163k1:sect163r1:sect163r2:sect193r1:sect193r2:sect233k1:sect233r1:sect239k1:sect283k1:sect283r1:sect409k1:sect409r1:sect571k1:sect571r1:secp160k1:secp160r1:secp160r2:secp192k1:prime192v1:secp224k1:secp224r1:secp256k1:prime256v1:secp384r1:secp521r1:brainpoolP256r1:brainpoolP384r1:brainpoolP512r1 [...]

But this does not, because it has 29 curves:
> $ openssl s_client -curves sect163k1:sect163r1:sect163r2:sect193r1:sect193r2:sect233k1:sect233r1:sect239k1:sect283k1:sect283r1:sect409k1:sect409r1:sect571k1:sect571r1:secp160k1:secp160r1:secp160r2:secp192k1:prime192v1:secp224k1:secp224r1:secp256k1:prime256v1:secp384r1:secp521r1:brainpoolP256r1:brainpoolP384r1:brainpoolP512r1:X25519 [...]

**Comment by kaduk** [comment]
It's not clear to me that competionism is a feature here; can you say more about why you want to use such a large curve list?

**Comment by tycho** [comment]
It's being used by automated tooling ([cipherscan](https://github.com/tycho/cipherscan)) which will use varying curve lists to see what the server accepts/prefers.

Why limit the list size to anything less than the number of available curves?

**Comment by richsalz** [comment]
Yes, we should set it to the number of curves that are available.  Should be a trivial fix.  Having a PR (hint hint) would help :)

**Comment by tycho** [comment]
@richsalz I'd be happy to make a PR, but I'm not sure how many curves are actually supported. I assumed something like this would be sufficient:
```
$ openssl ecparam -list_curves | grep : | wc -l
81
```
But that list doesn't include X25519 or X448 for some reason?

**Comment by mspncp** [comment]
`EC_get_builtin_curves(NULL, 0)` will return the number of curves, see implemtation of the `openssl ecparam -list_curves` command in [apps/ecparam.c, L180](https://github.com/openssl/openssl/blob/master/apps/ecparam.c#L180).

**Comment by tycho** [comment]
Right, but isn't `openssl ecparam -list_curves` an incomplete list? I see no mention of X25519 or (in git builds) X448 in that list, and I know those curves are supported.

**Comment by richsalz** [comment]
The list-curves flag does not list things that are used only for ECDH.  Once Ed25519 and 448 are present, *those* should show up.  Perhaps the list-curves doc should be improved?

**Comment by mspncp** [comment]
@mattcaswell, shouldn't X448 and Ed448 show up, now that TLS support (#5470) has been merged ? `openssl ecparam -list_curves  | grep 448` does not show any match

**Comment by mattcaswell** [comment]
`ecparam` is a tool for manipulating EC parameter files. Curve25519 and Curve448 are *special* and do not use EC parameter files (they do not work with standard ECDH(E) or ECSDA - instead they use X25519/X448 and Ed25519/Ed448). `-list_curves` only gives you the standard built-in curves that use EC parameter files.

**Comment by mattcaswell** [comment]
There are 30 built-ins (including the special curves). The complete list is here:

https://github.com/openssl/openssl/blob/master/ssl/t1_lib.c#L138

`MAX_CURVELIST` should be defined like this:

````
# define MAX_CURVELIST OSSL_NELEM(nid_list)
````
